dll/win32/iernonce/dialog.cpp

```diff
@@ -0,0 +1,220 @@
+
+HFONT CreateBoldFont(HFONT hOrigFont)
+{
+    LOGFONT fontAttributes = { 0 };
+    GetObject(hOrigFont, sizeof(fontAttributes), &fontAttributes);
+    fontAttributes.lfWeight = FW_BOLD;
+
+    return CreateFontIndirect(&fontAttributes);
+}
+
+ProgressDlg::ProgressDlg(RunOnceExInstance &RunOnceExInst) :
+    m_DialogID(-1), m_hListBox(NULL), m_hBoldFont(NULL), m_PointedItem(0), m_RunOnceExInst(RunOnceExInst)
+{ ; }
+
+BOOL ProgressDlg::RunDialogBox()
+{
+    // only when the list is not empty.
+    if (m_RunOnceExInst.m_SectionList.GetSize() != 0)
+    {
+        m_DialogID = DoModal();
+        if (m_DialogID == -1)
+        {
+            return FALSE;
+        }
+    }
+    return TRUE;
+}
+
+void ProgressDlg::CalcTextRect(LPCWSTR lpText, RECT *pRect)
+{
+    HDC hdc = ::GetDC(m_hListBox);
+    ::GetClientRect(m_hListBox, pRect);
+
+    pRect->bottom = pRect->top;
+    pRect->left += ITEM_LEFTPADDING;
+
+    HFONT OldFont = SelectFont(hdc, GetFont());
+
+    DrawTextW(hdc, lpText, -1, pRect, DT_CALCRECT | DT_WORDBREAK);
+    ::ReleaseDC(m_hListBox, hdc);
+
+    SelectFont(hdc, OldFont);
+
+    pRect->bottom -= pRect->top;
+    pRect->bottom += ITEM_VPADDING * 2;
+    pRect->top = 0;
+    pRect->right -= pRect->left;
+    pRect->left = 0;
+}
+
+void ProgressDlg::ResizeListBoxAndDialog(int NewHeight)
+{
+    RECT ListBoxRect;
+    RECT DlgRect;
+    ::GetWindowRect(m_hListBox, &ListBoxRect);
+    GetWindowRect(&DlgRect);
+
+    int HeightDiff = NewHeight - (ListBoxRect.bottom - ListBoxRect.top);
+
+    ::SetWindowPos(m_hListBox, NULL,
+        0, 0, ListBoxRect.right - ListBoxRect.left, NewHeight,
+        SWP_NOMOVE | SWP_NOZORDER | SWP_SHOWWINDOW);
+
+    SetWindowPos(0, 0, 0,
+        DlgRect.right - DlgRect.left, DlgRect.bottom - DlgRect.top + HeightDiff,
+        SWP_NOMOVE | SWP_NOZORDER | SWP_SHOWWINDOW);
+}
+
+BOOL
+ProgressDlg::ProcessWindowMessage(
+    HWND hwnd, UINT message, WPARAM wParam, LPARAM lParam, LRESULT &lResult,
+    DWORD dwMsgMapID)
+{
+    switch (message)
+    {
+        case WM_INITDIALOG:
+        {
+            if (!m_RunOnceExInst.m_Title.IsEmpty())
+            {
+                SetWindowTextW(m_RunOnceExInst.m_Title);
+            }
+
+            m_hListBox = GetDlgItem(IDC_LB_ITEMS);
+
+            m_hBoldFont = CreateBoldFont(GetFont());
+
+            // Add all sections with non-empty title into listbox
+            int TotalHeight = 0;
+            for (int i = 0; i < m_RunOnceExInst.m_SectionList.GetSize(); i++)
+            {
+                RunOnceExSection &Section = m_RunOnceExInst.m_SectionList[i];
+
+                if (!Section.m_SectionTitle.IsEmpty())
+                {
+                    INT Index = ListBox_AddString(m_hListBox, Section.m_SectionTitle);
+                    TotalHeight += ListBox_GetItemHeight(m_hListBox, Index);
+                    ListBox_SetItemData(m_hListBox, Index, i);
+                }
+            }
+
+            // Remove the WS_EX_CLIENTEDGE exstyle.
```

Comment:
Yeah well we all see that, you're paraphrasing what the code does just below.
If you want to add a useful comment, tell instead what kind of effect is done, when that style is removed.

Reply:
Oh, indeed it's obvious, I will try to avoid comments like this next time!